### PR TITLE
layout: Make the compositor rather than layout determine the position of each iframe.

### LINKED ISSUE #7377: Iframe content not painted at the right coordinate if the iframe is in a transformed element
Related: https://github.com/servo/servo/issues/7282

```html
<body style="background:red">
  <div style="transform: scale(1)">
    <iframe src="http://example.com"></iframe>
  </div>
</body>
```

screen shot 2015-08-26 at 10 51 15

### PR DESCRIPTION
**Pull request by pcwalton**
The old code that attempted to do this during layout wasn't able to work
for multiple reasons: it couldn't know where the iframe was going to be
on the page (because of nested iframes), and at the time it was building
the display list for a fragment it couldn't know where that fragment was
going to be in page coordinates.

This patch rewrites that code so that only the size of an iframe is
determined during layout, and the position is determined by the
compositor. Layout layerizes iframes and marks the iframe layers with
the appropriate subpage ID so that the compositor can place them
correctly.

Closes #7377.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7423)

### CODE REVIEW

File: components/compositing/compositor.rs [hunk 1]
```diff
@@ -644,9 +650,11 @@ impl<Window: WindowMethods> IOCompositor<Window> {
         // All root layers mask to bounds.
         *root_layer.masks_to_bounds.borrow_mut() = true;
 
-        if let Some(ref frame_rect) = frame_rect {
-            let frame_rect = frame_rect.to_untyped();
-            *root_layer.bounds.borrow_mut() = Rect::from_untyped(&frame_rect);
+        if let Some(ref frame_size) = frame_size {
+            let frame_size = frame_size.to_untyped();
+            root_layer.bounds.borrow_mut().size = Size2D::from_untyped(&frame_size);
+            /*root_layer.bounds.borrow_mut().origin =
```
Comment (glennw) [inline components/compositing/compositor.rs]: Can be removed

File: components/compositing/constellation.rs [hunk 2]
```diff
@@ -382,11 +382,17 @@ impl<LTF: LayoutTaskFactory, STF: ScriptTaskFactory> Constellation<LTF, STF> {
...
-            ConstellationMsg::FrameRect(pipeline_id, subpage_id, rect) => {
+            // A layout assigned a size to a subframe. This needs to be reflected by all frame
+            // trees in the navigation context containing the subframe.
+            ConstellationMsg::FrameSize(pipeline_id, subpage_id, size) => {
                 debug!("constellation got frame rect message");
-                self.handle_frame_rect_msg(pipeline_id, subpage_id, Rect::from_untyped(&rect));
+                self.handle_frame_size_msg(pipeline_id, subpage_id, &Size2D::from_untyped(&size));
+            }
+            ConstellationMsg::SetSubpagePosition(pipeline_id, subpage_id, position) => {
+                debug!("constellation got frame rect message");
```
Comment (glennw) [inline components/compositing/constellation.rs]: comment needs updating

File: components/layout/display_list_builder.rs [hunk 3]
```diff
@@ -1262,10 +1255,12 @@ impl FragmentDisplayListBuilding for Fragment {
         let layer = match layer {
             StackingContextLayer::Existing(existing_layer) => Some(existing_layer),
             StackingContextLayer::IfCanvas(layer_id) => {
```
Comment (glennw) [inline components/layout/display_list_builder.rs]: Perhaps rename ::IfCanvas() if it's now used for things other than canvas?

File: components/layout/display_list_builder.rs [hunk 4]
```diff
@@ -1814,21 +1818,28 @@ impl InlineFlowDisplayListBuilding for InlineFlow {
 
         // FIXME(Savago): fix Fragment::establishes_stacking_context() for absolute positioned item
         // and remove the check for filter presence. Further details on #5812.
-        has_stacking_context = has_stacking_context && {
-            if let SpecificFragmentInfo::Canvas(_) = self.fragments.fragments[0].specific {
-                true
-            } else {
-                !self.fragments.fragments[0].style().get_effects().filter.is_empty()
+        //
+        // FIXME(pcwalton): This is terribly, terribly bogus! What is even going on here?
```
Comment (glennw) [inline components/layout/display_list_builder.rs]: Can we add an issue for this - it looks very odd.

**Comment by glennw** [comment]
@pcwalton Just a few minor nits, looks good otherwise.

File: components/layout/display_list_builder.rs (hunk 4 above)
Comment (pcwalton) [inline components/layout/display_list_builder.rs]: Filed #7424.

**Comment by pcwalton** [comment]
Added some more expected test failures. I'm pretty sure the failures are pre-existing for the reasons stated in the commit message. Out of an abundance of caution I'd like to rerequest review. r? @glennw

**Comment by glennw** [comment]
@pcwalton Looking at some of those wpt/css failures, it's not immediately clear to me why they would start failing - could you provide some more info?

**Comment by pcwalton** [comment]
I fixed the failing test cases. Was a pretty simple fix to add `LAYERS_NEEDED_FOR_POSITIONED_DESCENDANTS`. Carrying forward review.

**Comment by jdm** [comment]
```

command timed out: 1200 seconds without output running ['./mach', 'test-ref', '--kind', 'cpu'], attempting to kill
```

**Comment by pcwalton** [comment]
I believe the races are fixed now.

r? @glennw

**Comment by glennw** [comment]
Reviewed 14 of 16 files at r1.
Review status: 0 of 23 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[components/compositing/constellation.rs, line 392 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7423#-Jxm24ce40HS_1sINEuj-r1-392)** ([raw file](https://github.com/servo/servo/blob/d4d87ed32cf04f908c2470ca2270483da869db4e/components/compositing/constellation.rs#L392)):</sup>
comment needs updating

---

<sup>**[components/layout/display_list_builder.rs, line 1257 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7423#-Jxm24ce40HS_1sINEuo-r1-1257)** ([raw file](https://github.com/servo/servo/blob/d4d87ed32cf04f908c2470ca2270483da869db4e/components/layout/display_list_builder.rs#L1257)):</sup>
Perhaps rename ::IfCanvas() if it's now used for things other than canvas?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7423)

**Comment by glennw** [comment]
Reviewed 2 of 16 files at r1, 18 of 21 files at r2.
Review status: 20 of 23 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/node.rs, line 248 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7423#-JzwzhipwPtyINceHszD)** ([raw file](https://github.com/servo/servo/blob/09b970af4aeba749146a9a9215e346653016672b/components/script/dom/node.rs#L248)):</sup>
Why is ReapLayoutData not needed anymore?

---

<sup>**[components/script/dom/window.rs, line 803 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7423#-JzwznFqyZqFKJyDqAmg)** ([raw file](https://github.com/servo/servo/blob/09b970af4aeba749146a9a9215e346653016672b/components/script/dom/window.rs#L803)):</sup>
If this comment can be removed, the GC should be too?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7423)

**Comment by glennw** [comment]
@pcwalton This generally looks OK to me (it seems to match what we discussed at the work week).

I'm not sure that ReapLayoutData can be removed though? There are a few nits apart from that question, and test-tidy needs to be fixed.

It's really unfortunate that the subpage abstraction is now leaking into the compositor, but I don't have any suggestion for a better way to fix it right now (other than fixing #5235 and #5241) :(

**Comment by pcwalton** [comment]
Review status: 20 of 23 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/node.rs, line 248 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7423#-JzwzhipwPtyINceHszD)** ([raw file](https://github.com/servo/servo/blob/09b970af4aeba749146a9a9215e346653016672b/components/script/dom/node.rs#L248)):</sup>
The `.unwrap()` line was causing trouble here because the layout task could have been shut down. It didn't seem like it was actually doing anything anymore, since all `handle_reap_layout_data` does is call `layout_data_wrapper.remove_compositor_layers` (which is now gone) so I just removed it.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7423)

**Comment by glennw** [comment]
Review status: 20 of 23 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/node.rs, line 248 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7423#-JzwzhipwPtyINceHszD)** ([raw file](https://github.com/servo/servo/blob/09b970af4aeba749146a9a9215e346653016672b/components/script/dom/node.rs#L248)):</sup>
There is the line "let layout_data_wrapper: LayoutDataWrapper = transmute(layout_data);" in handle_reap_layout_data() that does a transmute to a different type. I *think* this causes memory to be freed correctly, which may otherwise leak without reap layout data. Perhaps @jdm could comment on that?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7423)

**Comment by pcwalton** [comment]
@glennw New version of the patch restores the layout data reaping. Embarrassing that I forgot about that :)

**Comment by glennw** [comment]
@pcwalton Looks good to me once travis is happy (which I assume is a test-tidy failure but the travis results won't load for me for some reason).

**Comment by jdm** [comment]
```

/css-flexbox-1_dev/html/flexbox-basic-iframe-horiz-001.htm
----------------------------------------------------------
TIMEOUT expected FAIL [Parent]
/css-flexbox-1_dev/html/flexbox-basic-iframe-vert-001.htm
---------------------------------------------------------
TIMEOUT expected FAIL [Parent]
```

**Comment by pcwalton** [comment]
OK, I believe I've fixed those tests. The failures came from a deadlock in the way the compositor blocked on the constellation, which could happen simultaneously with the constellation waiting on the compositor. Fixing this by introducing a new message pair (`PrepareForSubpageLayerCreation` and `CreateLayerForSubpage`) exposed another race in detecting whether screenshots were safe to take, which I fixed by modifying the constellation; see the comment above `if let Some(size) = pipeline.size` in `constellation.rs`.

I also effectively redid #7771 with an alternate approach in order to fix the rebase conflict.

r? @glennw

**Comment by glennw** [comment]
Reviewed 3 of 21 files at r2, 12 of 12 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7423)

**Comment by mbrubeck** [comment]
@bors-servo retry

#7787

**Pull request merged**